apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,483 @@
+import concurrent.futures
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ContextManager,
+    Dict,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+import anndata
+import attrs
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+from typing_extensions import Literal, TypedDict
+
+from .. import somacore  # to be replaced by somacore package, when available
+from ..dataframe import DataFrame as SOMADataFrame
+from ..sparse_nd_array import SparseNDArrayRead
+
+if TYPE_CHECKING:
+    from ..experiment import Experiment
+from ..sparse_nd_array import SparseNDArray as SOMASparseNDArray
+from .anndata import _make_anndata
+from .axis import AxisQuery
+from .eq_types import AxisColumnNames, ExperimentAxisQueryReadArrowResult
+
+AxisJoinIds = TypedDict(
+    "AxisJoinIds",
+    {
+        "obs": Optional[pa.Array],
+        "var": Optional[pa.Array],
+    },
+)
+
+MatrixAxisQuery = TypedDict(
+    "MatrixAxisQuery",
+    {
+        "obs": AxisQuery,
+        "var": AxisQuery,
+    },
+)
+
+
+class ExperimentAxisQuery(ContextManager["ExperimentAxisQuery"]):
```

Comment:
You could use [`contextlib.AbstractContextManager`](https://docs.python.org/3/library/contextlib.html) as the base here and avoid needing to implement your own `__enter__` method.

Reply:
I am unable to get AbstractContextManager to work reliably. On older versions of Python, it is not a subscriptable (generic), but mypy is failing when I omit the parameter.

For now, reverting to the contextlib.ContextManager, which seems to work correctly across all versions, albeit requiring an `__enter__` implementation.